### PR TITLE
PodSecurity message/check/fixture cleanups

### LINKED ISSUE #103373: [PodSecurity] make name/reason/details messages consistent across checks
Once all checks are implemented (https://github.com/kubernetes/kubernetes/issues/103194), do a sweep to ensure the forbidden reason and details are consistent/coherent across them

* Consistency of check IDs
* Make reason consistent (and consider exporting as a constant)
* Make details consistent
  * e.g. https://github.com/kubernetes/kubernetes/pull/103382#discussion_r662412946
* Ensure fixtures balance count/coverage (especially with variants covering initContainers)
* Avoid duplicate messages from overlapping checks (baseline hostPath and restricted volumes)
* Comment blocks capturing the current policy (take it or leave it, but should be consistent)
* Usage of sets over slices (sets shuffle the order, and are less efficient when de-duplication isn't needed)

/assign

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind cleanup
/kind feature

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/103373

Best reviewed commit-by-commit. Alternate commits are updating generated fixtures.

Address the following:
* Consistency of check IDs
* Make reason consistent (and consider exporting as a constant)
* Make details consistent
  * e.g. https://github.com/kubernetes/kubernetes/pull/103382#discussion_r662412946
* Ensure fixtures balance count/coverage (especially with variants covering initContainers)
  * Reduced from 3966 to 2553
* Comment blocks capturing the current policy (take it or leave it, but should be consistent)
* Usage of sets over slices (sets shuffle the order, and are less efficient when de-duplication isn't needed)

---

Review pass 1 (https://github.com/kubernetes/kubernetes/pull/103558#pullrequestreview-701536894):

checks:
- [x] allowPrivilegeEscalation
- [x] appArmorProfile
- [x] addCapabilities → capabilities_baseline
- [x] dropCapabilities → capabilities_restricted
- [x] hostNamespaces
- [x] hostPath → hostPathVolumes
- [x] hostPorts

---

Unreviewed:

checks:
- [x] hostProcess → windowsHostProcess
- [x] privileged
- [x] procMount
- [x] restrictedVolumes
- [x] runAsNonRoot
- [x] seccomp_restricted → seccompProfile_restricted
- [x] seccomp_baseline → seccompProfile_baseline
- [x] sysctls
- [x] selinux → seLinuxOptions

cleanup: drop field path from container visitor

limit warning evaluation/surfacing to requests that are not rejected by an enforce policy

reformat the warning message to read more naturally when admitting pods or pod-controllers

---

Follow-up:

* Avoid duplicate messages from overlapping checks (tracked in https://github.com/kubernetes/kubernetes/issues/103561)
  * restricted volumes complains about hostPath volumes already caught by baseline hostPath check
  * restricted capabilities complains about added non-default capabilities already caught by baseline capabilities check
  * restricted seccomp complains about unconfined seccomp profile already caught by baseline seccomp check

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @tallclair

### CODE REVIEW

File: staging/src/k8s.io/pod-security-admission/policy/check_capabilities_restricted.go [hunk 1]
```diff
@@ -64,36 +83,39 @@ func capabilitiesRestricted_1_22(podMetadata *metav1.ObjectMeta, podSpec *corev1
...
+		}
 	})
 	var forbiddenDetails []string
 	if len(containersMissingDropAll) > 0 {
-		forbiddenDetails = append(forbiddenDetails, fmt.Sprintf("containers %q must drop ALL", containersMissingDropAll))
+		forbiddenDetails = append(forbiddenDetails, fmt.Sprintf(`containers %s must set securityContext.capabilities.drop=["ALL"]`, joinQuote(containersMissingDropAll)))
 	}
 	if len(containersAddingForbidden) > 0 {
-		forbiddenDetails = append(forbiddenDetails, fmt.Sprintf("containers %q must not add %q", containersAddingForbidden, forbiddenCapabilities.List()))
+		forbiddenDetails = append(forbiddenDetails, fmt.Sprintf(`containers %s must not include %s in securityContext.capabilities.add`, joinQuote(containersAddingForbidden), joinQuote(forbiddenCapabilities.List())))
```
Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/policy/check_capabilities_restricted.go:113]: nit: make this message consistent with the baseline capabilities details.

File: staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go [hunk 2]
```diff
@@ -56,31 +56,30 @@ func CheckAppArmorProfile() Check {
...
 	}
-
 	if len(forbiddenValues) > 0 {
+		sort.Strings(forbiddenValues)
+		var reason string
+		if len(forbiddenValues) == 1 {
+			reason = "forbidden AppArmor profile"
+		} else {
+			reason = "forbidden AppArmor profiles"
+		}
```
Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:78]: nit: use pluralize

**Review by tallclair — COMMENTED** [review]
lgtm so far

File: staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:78]: will do... didn't exist when I wrote this one  :)

File: staging/src/k8s.io/pod-security-admission/policy/check_restrictedVolumes.go [hunk 3]
```diff
@@ -89,16 +98,71 @@ func restrictedVolumes_1_0(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSp
...
+				badVolumeTypes.Insert("photonPersistentDisk")
+			case volume.PortworxVolume != nil:
+				badVolumeTypes.Insert("portworxVolume")
+			case volume.ScaleIO != nil:
+				badVolumeTypes.Insert("scaleIO")
+			case volume.StorageOS != nil:
+				badVolumeTypes.Insert("storageos")
+			default:
+				badVolumeTypes.Insert("unknown")
+			}
```
Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/policy/check_restrictedVolumes.go:150]: I suppose we shouldn't depend on PSP code? We might want to move https://github.com/kubernetes/kubernetes/blob/24a614cb4c5d8eaedffb4ad3b77e3c000d6e2b79/pkg/security/podsecuritypolicy/util/util.go#L81 to a common location.

**Review by tallclair — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/pod-security-admission/policy/check_restrictedVolumes.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/pod-security-admission/policy/check_restrictedVolumes.go:150]: I don't want to add any new dependencies on that code... looking forward to deleting it in 1.25

**Pull request merged**